### PR TITLE
Fixed some remote play commands

### PR DESCRIPTION
**Pull request by mueslimak3r**
**Changes**

* `TVApiEventListener` runs player interaction on the main thread. (there is probably a better way to run things on the main thread). If the interactions do not happen on the same thread as the player it will throw an exception:
`SimpleExoPlayer: Player is accessed on the wrong thread`

* updated the remote playstate commands to interact with audio and video sessions appropriately

* reworked `mediaManager` audio queue and reporting system to accommodate switching between audio and video sessions

* audio player is released when stopped remotely or by switching to playing video
 
**Issues**
* remote playstate commands were mostly disfunctional
* switching between audio and video playback caused double reporting and didn't release audio player

### CODE REVIEW

File: app/src/main/java/org/jellyfin/androidtv/data/eventhandling/TvApiEventListener.java [hunk 1]
```diff
@@ -74,11 +75,15 @@ public void onLibraryChanged(ApiClient client, LibraryUpdateInfo info) {
     @Override
     public void onPlaystateCommand(ApiClient client, PlaystateRequest command) {
         PlaybackController playbackController = TvApp.getApplication().getPlaybackController();
+        Timber.d("caught playstate command: %s", command.getCommand());
+
+        //handler to access the players on the main thread
+        Handler mainThreadHandler = new Handler(Looper.getMainLooper());
```
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/data/eventhandling/TvApiEventListener.java:81]: It's probably more efficient to create one handler as class property and re-use that one

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]
For the next time; please split changes into multiple PR's. I assumed this PR would only change the WebSocket code with some small tweaks in MediaManager but it contains other changes in the MediaManager like the progress reporting functionality.

**Comment by mueslimak3r** [comment]
I made the handler a class property which is much tidier.

> For the next time; please split changes into multiple PR's. I assumed this PR would only change the WebSocket code with some small tweaks in MediaManager but it contains other changes in the MediaManager like the progress reporting functionality.

Regarding separating the changes into multiple PRs, using this one as an example would it make sense to:

1) make a PR for all the absolutely necessary `mediaManager` changes to allow interfacing with the api event listener, including the reporting changes (these prevent the remote from getting confused about what's playing and flashing between the stopped audio item and the playing video item)

2) if the prior PR is approved, create a PR just for the TVApiEventListener changes

File: app/src/main/java/org/jellyfin/androidtv/data/eventhandling/TvApiEventListener.java (hunk 1 above)
Comment (mueslimak3r) [reply app/src/main/java/org/jellyfin/androidtv/data/eventhandling/TvApiEventListener.java]: done

**Comment by nielsvanvelzen** [comment]
> Regarding separating the changes into multiple PRs, using this one as an example would it make sense to:
> 
> 1. make a PR for all the absolutely necessary `mediaManager` changes to allow interfacing with the api event listener, including the reporting changes (these prevent the remote from getting confused about what's playing and flashing between the stopped audio item and the playing video item)
> 2. if the prior PR is approved, create a PR just for the TVApiEventListener changes

Yeah pretty much, you'll separate unrelated changes to separate pull requests. You can always create all of them at once and add something like "depends on #1273" in the pr descriptions.

It's just way easier to test and review when the PR is smaller and the focus is on a single change instead of a bunch of them.

File: app/src/main/java/org/jellyfin/androidtv/data/eventhandling/TvApiEventListener.java [hunk 2]
```diff
@@ -39,10 +40,14 @@
...
     private final DataRefreshService dataRefreshService;
     private final MediaManager mediaManager;
+    private final Handler mainThreadHandler;
 
     public TvApiEventListener(DataRefreshService dataRefreshService, MediaManager mediaManager) {
         this.dataRefreshService = dataRefreshService;
         this.mediaManager = mediaManager;
+
+        //handler to interact with the players (exoplayer doesn't allow access from another thread)
+        this.mainThreadHandler = new Handler(Looper.getMainLooper());
```
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/data/eventhandling/TvApiEventListener.java:50]: Don't need `this.`:

```suggestion
        mainThreadHandler = new Handler(Looper.getMainLooper());
```

**Review by nielsvanvelzen — APPROVED** [review]
Changes look fine but I experiences a crash while testing, I lost the logs of that so I'm still trying to reproduce it.

**Comment by nielsvanvelzen** [comment]
When I start the playback of an item via the remote control and it ends the app crashes. I don't think it's caused by this pull request though.

```
    --------- beginning of crash
2021-12-08 21:36:42.016 7477-7477/org.jellyfin.androidtv.debug E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.jellyfin.androidtv.debug, PID: 7477
    java.lang.UnsupportedOperationException
        at java.util.AbstractList.remove(AbstractList.java:167)
        at org.jellyfin.androidtv.ui.playback.PlaybackController.removePreviousQueueItems(PlaybackController.java:1153)
        at org.jellyfin.androidtv.ui.playback.CustomPlaybackOverlayFragment.showNextUp(CustomPlaybackOverlayFragment.java:1378)
        at org.jellyfin.androidtv.ui.playback.PlaybackController.itemComplete(PlaybackController.java:1176)
        at org.jellyfin.androidtv.ui.playback.PlaybackController.access$3500(PlaybackController.java:61)
        at org.jellyfin.androidtv.ui.playback.PlaybackController$17.onEvent(PlaybackController.java:1296)
        at org.jellyfin.androidtv.ui.playback.VideoManager$2.onPlayerStateChanged(VideoManager.java:123)
        at com.google.android.exoplayer2.ExoPlayerImpl.lambda$updatePlaybackInfo$15(ExoPlayerImpl.java:1303)
        at com.google.android.exoplayer2.ExoPlayerImpl$$ExternalSyntheticLambda1.invoke(Unknown Source:4)
        at com.google.android.exoplayer2.util.ListenerSet$ListenerHolder.invoke(ListenerSet.java:283)
        at com.google.android.exoplayer2.util.ListenerSet.lambda$queueEvent$0(ListenerSet.java:170)
        at com.google.android.exoplayer2.util.ListenerSet$$ExternalSyntheticLambda1.run(Unknown Source:6)
        at com.google.android.exoplayer2.util.ListenerSet.flushEvents(ListenerSet.java:191)
        at com.google.android.exoplayer2.ExoPlayerImpl.updatePlaybackInfo(ExoPlayerImpl.java:1340)
        at com.google.android.exoplayer2.ExoPlayerImpl.handlePlaybackInfo(ExoPlayerImpl.java:1175)
        at com.google.android.exoplayer2.ExoPlayerImpl.lambda$new$1$ExoPlayerImpl(ExoPlayerImpl.java:216)
        at com.google.android.exoplayer2.ExoPlayerImpl$$ExternalSyntheticLambda14.run(Unknown Source:4)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
2021-12-08 21:36:42.017 7477-7477/org.jellyfin.androidtv.debug E/ACRA: ACRA caught a UnsupportedOperationException for org.jellyfin.androidtv.debug
    java.lang.UnsupportedOperationException
        at java.util.AbstractList.remove(AbstractList.java:167)
        at org.jellyfin.androidtv.ui.playback.PlaybackController.removePreviousQueueItems(PlaybackController.java:1153)
        at org.jellyfin.androidtv.ui.playback.CustomPlaybackOverlayFragment.showNextUp(CustomPlaybackOverlayFragment.java:1378)
        at org.jellyfin.androidtv.ui.playback.PlaybackController.itemComplete(PlaybackController.java:1176)
        at org.jellyfin.androidtv.ui.playback.PlaybackController.access$3500(PlaybackController.java:61)
        at org.jellyfin.androidtv.ui.playback.PlaybackController$17.onEvent(PlaybackController.java:1296)
        at org.jellyfin.androidtv.ui.playback.VideoManager$2.onPlayerStateChanged(VideoManager.java:123)
        at com.google.android.exoplayer2.ExoPlayerImpl.lambda$updatePlaybackInfo$15(ExoPlayerImpl.java:1303)
        at com.google.android.exoplayer2.ExoPlayerImpl$$ExternalSyntheticLambda1.invoke(Unknown Source:4)
        at com.google.android.exoplayer2.util.ListenerSet$ListenerHolder.invoke(ListenerSet.java:283)
        at com.google.android.exoplayer2.util.ListenerSet.lambda$queueEvent$0(ListenerSet.java:170)
        at com.google.android.exoplayer2.util.ListenerSet$$ExternalSyntheticLambda1.run(Unknown Source:6)
        at com.google.android.exoplayer2.util.ListenerSet.flushEvents(ListenerSet.java:191)
        at com.google.android.exoplayer2.ExoPlayerImpl.updatePlaybackInfo(ExoPlayerImpl.java:1340)
        at com.google.android.exoplayer2.ExoPlayerImpl.handlePlaybackInfo(ExoPlayerImpl.java:1175)
        at com.google.android.exoplayer2.ExoPlayerImpl.lambda$new$1$ExoPlayerImpl(ExoPlayerImpl.java:216)
        at com.google.android.exoplayer2.ExoPlayerImpl$$ExternalSyntheticLambda14.run(Unknown Source:4)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

**Comment by mueslimak3r** [comment]
I fixed the crash. `TvApiEventListener` was feeding an `Arrays.asList()` which is fixed size to `mediaManager.setCurrentVideoQueue()`. The fix is to make a new list from the input. The audio queue already handles this.

While testing that I also found an issue where the onComplete -> nextup -> new playback session was just playing the same episode over again. doing the queue item removal to `mediaManager`'s queue fixed it. I'm not sure why that is needed.

In the future I'd like to have all queue related modifications be done via methods within mediaManager so they can be built to handle all the weird cases. As it is now, there's kinda a mess of special case handlers in various places external to mediaManager, and it would be easy to accidentally not account for them.

File: app/src/main/java/org/jellyfin/androidtv/data/eventhandling/TvApiEventListener.java (hunk 2 above)
Comment (mueslimak3r) [reply app/src/main/java/org/jellyfin/androidtv/data/eventhandling/TvApiEventListener.java]: done

**Review by nielsvanvelzen — APPROVED** [review]

**Pull request merged**